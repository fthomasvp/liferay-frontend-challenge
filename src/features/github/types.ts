export type TGitHubRepo = {
  id: number;
  full_name: string;
  created_at: string;
  stargazers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: {
    name: string;
  } | null;
  lastCommitAt: string | null;
  isFavorited: boolean;
};

export type TGitHubCommit = {
  commit: {
    committer: {
      date: string;
      email: string;
      name: string;
    };
  };
};

export type TGitHubParams = {
  username: string;
  repositoryName: string;
};
