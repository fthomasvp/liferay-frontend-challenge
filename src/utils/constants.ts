export const MINIMUM_REPO_NAME_LENGTH = 1;
