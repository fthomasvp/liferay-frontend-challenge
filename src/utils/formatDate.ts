import { formatDistanceToNow } from 'date-fns';

export const formatDistanceTimeFromNow = (date: string) => {
  return formatDistanceToNow(new Date(date), {
    addSuffix: true,
  });
};
