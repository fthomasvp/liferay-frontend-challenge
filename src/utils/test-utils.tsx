import React, { ProviderProps, ReactElement } from 'react';
import { render } from '@testing-library/react';

import { HomeContext, THomeContext } from 'context/HomeContext';
import {
  SearchBarContext,
  SearchBarContextType,
} from 'context/search-bar/search-bar.context';

/**
 * A custom render to setup providers. Extends regular
 * render options with `providerProps` to allow injecting
 * different scenarios to test with.
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
const customRender = (
  ui: ReactElement,
  {
    homeProviderProps,
    searchBarProviderProps,
    ...renderOptions
  }: {
    homeProviderProps: ProviderProps<THomeContext>;
    searchBarProviderProps: ProviderProps<SearchBarContextType>;
  }
) => {
  return render(
    <HomeContext.Provider {...homeProviderProps}>
      <SearchBarContext.Provider {...searchBarProviderProps}>
        {ui}
      </SearchBarContext.Provider>
    </HomeContext.Provider>,
    renderOptions
  );
};

export * from '@testing-library/react';
export { customRender as render };
