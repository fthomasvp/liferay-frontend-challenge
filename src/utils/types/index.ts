export type GitHubRepo = {
  id: number;
  full_name: string;
  created_at: string;
  stargazers_count: number;
  language: string;
  forks_count: number;
  open_issues_count: number;
  license: {
    name: string;
  } | null;
  lastCommitAt: string | null;
  isFavorited: boolean;
};

export type Filter = {
  repositoryName?: string;
  isFavorited?: boolean;
};
