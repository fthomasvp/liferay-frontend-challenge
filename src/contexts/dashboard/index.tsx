import React from 'react';

import { GitHubRepo, Filter } from 'utils/types';

type DashboardContextType = {
  repositories: GitHubRepo[];
  filteredRepositories?: GitHubRepo[];
  setFilteredRepositories: (repositories: GitHubRepo[]) => void;
  selectedRepository?: GitHubRepo;
  setSelectedRepository: (repository: GitHubRepo) => void;
  addRepository: (repository: GitHubRepo) => void;
  deleteRepository: () => void;
  filterRepositories: (filter: Filter) => void;
  orderRepositories: (e: React.MouseEvent<HTMLElement, MouseEvent>) => void;
  searchText: string;
  setSearchText: (text: string) => void;
  isFiltering: boolean;
  setIsFiltering: (isFiltering: boolean) => void;
  favorRepository: (repository: GitHubRepo) => void;
  starIcon: boolean;
  setStarIcon: (hasStarIcon: any) => void;
};

export const DashboardContext = React.createContext<DashboardContextType>({
  repositories: [],
  filteredRepositories: [],
  setFilteredRepositories: () => null,
  selectedRepository: {
    id: 1,
    full_name: '',
    created_at: '',
    stargazers_count: 1,
    language: '',
    forks_count: 1,
    open_issues_count: 1,
    license: {
      name: '',
    },
    lastCommitAt: '',
    isFavored: false,
  },
  setSelectedRepository: () => null,
  addRepository: () => null,
  deleteRepository: () => null,
  filterRepositories: () => null,
  orderRepositories: () => null,
  searchText: '',
  setSearchText: () => null,
  isFiltering: false,
  setIsFiltering: () => null,
  favorRepository: () => null,
  starIcon: true,
  setStarIcon: () => null,
});
