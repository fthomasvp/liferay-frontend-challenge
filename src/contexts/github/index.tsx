import React from 'react';

import { GitHubRepo } from 'utils/types';

type DashboardContext = {
  repositories: GitHubRepo[];
  addRepository: (repository: GitHubRepo) => void;
  selectedRepository?: GitHubRepo;
  setSelectedRepository: (repository: GitHubRepo) => void;
  deleteRepository: () => void;
};

export const GitHubContext = React.createContext<DashboardContext>({
  repositories: [],
  addRepository: () => console.log('Just for start'),
  selectedRepository: {
    id: 1,
    full_name: '',
    created_at: '',
    stargazers_count: 1,
    language: '',
    forks_count: 1,
    open_issues_count: 1,
    license: {
      name: '',
    },
    lastCommitAt: '',
  },
  setSelectedRepository: () => console.log('Just for start'),
  deleteRepository: () => console.log('Just for start'),
});
