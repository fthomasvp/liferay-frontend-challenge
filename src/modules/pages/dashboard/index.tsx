import React from 'react';
import _orderBy from 'lodash.orderby';

import Header from 'modules/components/header';
import DynamicDisplay from 'modules/components/dynamic-display';
import { DashboardContext } from 'contexts/dashboard';
import { Filter, GitHubRepo } from 'utils/types';

const Dashboard = (): JSX.Element => {
  const [repositories, setRepositories] = React.useState<GitHubRepo[]>([]);
  const [
    selectedRepository,
    setSelectedRepository,
  ] = React.useState<GitHubRepo>();
  const [filteredRepositories, setFilteredRepositories] = React.useState<
    GitHubRepo[]
  >([]);
  const [searchText, setSearchText] = React.useState('');
  const [isFiltering, setIsFiltering] = React.useState(false);
  const [starIcon, setStarIcon] = React.useState(true);

  const addRepository = React.useCallback(
    (repository) => {
      const newRepositories = [repository, ...repositories];

      setRepositories(newRepositories);
    },
    [repositories]
  );

  const deleteRepository = () => {
    if (isFiltering) {
      const newRepositories = filteredRepositories.filter(
        (repo) => repo.id !== selectedRepository?.id
      );

      setFilteredRepositories(newRepositories);
    }

    const newRepositories = repositories.filter(
      (repo) => repo.id !== selectedRepository?.id
    );

    setRepositories(newRepositories);
  };

  const orderRepositories = (e: React.MouseEvent<HTMLElement, MouseEvent>) => {
    const fieldToSort = (e.target as HTMLButtonElement).name;

    if (isFiltering) {
      setFilteredRepositories(
        _orderBy(filteredRepositories, [fieldToSort], ['desc']) as GitHubRepo[]
      );
    }

    setRepositories(
      _orderBy(repositories, [fieldToSort], ['desc']) as GitHubRepo[]
    );
  };

  const filterRepositories = (filter: Filter) => {
    if (repositories.length > 0) {
      let newFilteredRepositories: GitHubRepo[] = [];

      if (filter?.repositoryName) {
        newFilteredRepositories = repositories.filter((repo) => {
          if (!filter.repositoryName) {
            return false;
          }

          return repo.full_name.match(new RegExp(filter.repositoryName, 'i'));
        });
      }

      if (filter?.isFavored) {
        newFilteredRepositories = repositories.filter(
          (repo) => repo.isFavored === filter.isFavored
        );
      }

      setFilteredRepositories(newFilteredRepositories);
    }
  };

  const favorRepository = (repository: GitHubRepo) => {
    if (isFiltering) {
      const newRepositories = filteredRepositories.map((repo) => {
        if (repo.id === repository?.id) {
          repo.isFavored = !repository.isFavored;
        }

        return repo;
      });

      setFilteredRepositories(newRepositories);
    }

    const newRepositories = repositories.map((repo) => {
      if (repo.id === repository?.id) {
        repo.isFavored = !repository.isFavored;
      }

      return repo;
    });

    setRepositories(newRepositories);
  };

  return (
    <DashboardContext.Provider
      value={{
        repositories,
        filteredRepositories,
        setFilteredRepositories,
        selectedRepository,
        setSelectedRepository,
        addRepository,
        deleteRepository,
        filterRepositories,
        orderRepositories,
        searchText,
        setSearchText,
        isFiltering,
        setIsFiltering,
        favorRepository,
        starIcon,
        setStarIcon,
      }}
    >
      <Header />

      <DynamicDisplay
        isFilteredItems={isFiltering}
        items={isFiltering ? filteredRepositories : repositories}
      />
    </DashboardContext.Provider>
  );
};

export default Dashboard;
