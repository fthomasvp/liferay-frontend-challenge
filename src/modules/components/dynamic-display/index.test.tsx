import React from 'react';
import { render, cleanup } from '@testing-library/react';

import DynamicDisplay from './index';
import { DashboardContext } from 'contexts/dashboard';
import { GitHubRepo } from 'utils/types';

describe('<DynamicDisplay />', () => {
  afterEach(cleanup);

  const repositories: GitHubRepo[] = [
    {
      id: 22377139,
      full_name: 'liferay/clay',
      created_at: '2014-07-29T11:07:36Z',
      stargazers_count: 156,
      language: 'HTML',
      forks_count: 185,
      open_issues_count: 55,
      license: {
        name: 'Other',
      },
      lastCommitAt: '2021-03-05T12:23:22Z',
    },
  ];
  const addRepository = jest.fn();

  const props = {
    repositories,
    filteredRepositories: [],
    setFilteredRepositories: jest.fn(),
    selectedRepository: repositories[0],
    setSelectedRepository: jest.fn(),
    addRepository,
    deleteRepository: jest.fn(),
    filterRepositories: jest.fn(),
    orderRepositories: jest.fn(),
    searchText: '',
    setSearchText: jest.fn(),
    isFiltering: false,
    setIsFiltering: jest.fn(),
  };

  const dynamicDisplayProps = {
    items: repositories,
  };

  test('should render correctly', () => {
    const { container } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should display repository full name on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('fullName')).toHaveTextContent('liferay/clay');
  });

  test('should display repository stars count on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('stargazersCount')).toHaveTextContent('Stars 156');
  });

  test('should display repository forks count on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('forksCount')).toHaveTextContent('Forks 185');
  });

  test('should display repository open issues count on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('openIssues')).toHaveTextContent('Open issues 55');
  });

  test('should display repository age on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('age')).toBeInTheDocument();
  });

  test('should display repository last commit on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('lastCommitAt')).not.toHaveTextContent('No commits yet');
  });

  test('when no commit is found, should display "No commits yet" on card', () => {
    const [repository] = repositories;

    repository.lastCommitAt = null;

    const updatedProps = {
      ...props,
      repositories: [repository],
    };

    const { getByTestId } = render(
      <DashboardContext.Provider value={updatedProps}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('lastCommitAt')).toHaveTextContent('No commits yet');
  });

  test('should display repository license name on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('license')).toHaveTextContent('License Other');
  });

  test('when no license is found, should display "N/A" on card', () => {
    const [repository] = repositories;

    repository.license = null;

    const updatedProps = {
      ...props,
      repositories: [repository],
    };

    const { getByTestId } = render(
      <DashboardContext.Provider value={updatedProps}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('license')).toHaveTextContent('License N/A');
  });

  test('should display repository language label on card', () => {
    const { getByTestId } = render(
      <DashboardContext.Provider value={props}>
        <DynamicDisplay {...dynamicDisplayProps} />
      </DashboardContext.Provider>
    );
    expect(getByTestId('language')).toHaveTextContent('HTML');
  });
});
