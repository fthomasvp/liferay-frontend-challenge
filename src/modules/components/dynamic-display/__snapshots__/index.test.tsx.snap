// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DynamicDisplay /> should render correctly 1`] = `
<div>
  <div
    class="container-fluid container-view container-fluid-max-xl"
  >
    <div
      class="row"
    >
      <div
        class="col-md-4"
      >
        <div
          class="card card-type-asset"
        >
          <div
            class="card-row"
            style="height: 50px; align-items: center;"
          >
            <img
              alt="Liferay logo"
              src="liferay_logo.svg"
              style="margin-left: 6px;"
            />
            <div
              class="col-8"
            >
              <p
                class="card-title"
                data-testid="fullName"
                title="liferay/clay"
              >
                <span
                  class="text-truncate-inline"
                >
                  <span
                    class="text-truncate"
                  >
                    liferay/clay
                  </span>
                </span>
              </p>
            </div>
            <button
              aria-label="Star"
              class="btn btn-monospaced btn-unstyled"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-star-o"
                role="presentation"
              >
                <use
                  xlink:href="#star-o"
                />
              </svg>
            </button>
            <button
              aria-label="Trash"
              class="btn btn-monospaced btn-unstyled"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-trash"
                role="presentation"
              >
                <use
                  xlink:href="#trash"
                />
              </svg>
            </button>
          </div>
          <div
            class="card-divider"
            style="margin-top: 0px; margin-bottom: 0px;"
          />
          <div
            class="card-body"
          >
            <div
              class="card-row"
            >
              <div
                class="autofit-col autofit-col-expand"
              >
                <section
                  class="autofit-section"
                >
                  <p
                    class="card-title"
                    data-testid="stargazersCount"
                    title="Stars, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        Stars
                         
                        <span
                          class="font-weight-normal"
                        >
                          156
                        </span>
                      </span>
                    </span>
                  </p>
                  <p
                    class="card-title"
                    data-testid="forksCount"
                    title="Forks, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        Forks
                         
                        <span
                          class="font-weight-normal"
                        >
                          185
                        </span>
                      </span>
                    </span>
                  </p>
                  <p
                    class="card-title"
                    data-testid="openIssues"
                    title="Open issues, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        Open issues
                         
                        <span
                          class="font-weight-normal"
                        >
                          55
                        </span>
                      </span>
                    </span>
                  </p>
                  <p
                    class="card-title"
                    data-testid="age"
                    title="Age, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        Age
                         
                        <span
                          class="font-weight-normal"
                        >
                          over 6 years ago
                        </span>
                      </span>
                    </span>
                  </p>
                  <p
                    class="card-title"
                    data-testid="lastCommitAt"
                    title="Last commit, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        Last commit
                         
                        <span
                          class="font-weight-normal"
                        >
                          4 days ago
                        </span>
                      </span>
                    </span>
                  </p>
                  <p
                    class="card-title"
                    data-testid="license"
                    title="License, ,[object Object]"
                  >
                    <span
                      class="text-truncate-inline"
                    >
                      <span
                        class="text-truncate"
                      >
                        License
                         
                        <span
                          class="font-weight-normal"
                        >
                          Other
                        </span>
                      </span>
                    </span>
                  </p>
                  <div
                    class="card-detail"
                  >
                    <span
                      class="label label-danger"
                      data-testid="language"
                    >
                      <span
                        class="label-item label-item-expand"
                      >
                        HTML
                      </span>
                    </span>
                  </div>
                </section>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
